File: src/tinohelm/cli/_style.py
"""Shared CLI styling utilities for consistent, ergonomic terminal output."""
from __future__ import annotations

from typing import Any

import typer


# ── Color Palette ──────────────────────────────────────────────────────────
# Semantic color mapping for consistent theming across all CLI commands.

class C:
    """Semantic color constants."""
    POSITIVE = typer.colors.GREEN
    NEGATIVE = typer.colors.RED
    NEUTRAL = typer.colors.WHITE
    ACCENT = typer.colors.CYAN
    WARN = typer.colors.YELLOW
    MUTED = typer.colors.BRIGHT_BLACK
    HIGHLIGHT = typer.colors.MAGENTA

    # Status-specific
    STATUS = {
        "completed": typer.colors.GREEN,
        "running": typer.colors.YELLOW,
        "queued": typer.colors.CYAN,
        "failed": typer.colors.RED,
        "cancelled": typer.colors.MAGENTA,
        "error": typer.colors.RED,
    }

    STATUS_ICON = {
        "completed": "+",
        "running": "~",
        "queued": ".",
        "failed": "x",
        "cancelled": "!",
        "error": "x",
    }


# ── Text Helpers ───────────────────────────────────────────────────────────

def bold(text: str) -> str:
    return typer.style(str(text), bold=True)


def dim(text: str) -> str:
    return typer.style(str(text), dim=True)


def accent(text: str) -> str:
    return typer.style(str(text), fg=C.ACCENT)


def muted(text: str) -> str:
    return typer.style(str(text), fg=C.MUTED)


def color_value(value: float | None, fmt: str = "+.2f") -> str:
    """Format and color a numeric value: green if positive, red if negative."""
    if value is None:
        return muted("-")
    s = f"{value:{fmt}}"
    if value > 0:
        return typer.style(s, fg=C.POSITIVE)
    elif value < 0:
        return typer.style(s, fg=C.NEGATIVE)
    return s


def color_status(status: str) -> str:
    """Color a status string."""
    fg = C.STATUS.get(status, C.NEUTRAL)
    return typer.style(status, fg=fg, bold=True)


def status_badge(status: str) -> str:
    """Render a colored [icon] badge for a status."""
    fg = C.STATUS.get(status, C.NEUTRAL)
    icon = C.STATUS_ICON.get(status, "?")
    return typer.style(f"[{icon}]", fg=fg)


# ── Layout Helpers ─────────────────────────────────────────────────────────

def header(title: str) -> None:
    """Print a section header with divider."""
    typer.echo()
    typer.echo(f"  {bold(title)}")


def divider(width: int = 70) -> None:
    typer.echo("  " + "-" * width)


def kv(label: str, value: str, label_width: int = 20) -> None:
    """Print a key-value pair with right-aligned label."""
    typer.echo(f"    {label:>{label_width}s}: {value}")


def kv_color(label: str, value: float | None, fmt: str = ".4f",
             label_width: int = 20, colorize: bool = False,
             pct: bool = False, pct_x100: bool = False) -> None:
    """Print a key-value pair with optional coloring and percentage formatting."""
    if value is None:
        kv(label, muted("-"), label_width)
        return

    if pct_x100:
        v_str = f"{value * 100:{fmt}}%"
    elif pct:
        v_str = f"{value:{fmt}}%"
    else:
        v_str = f"{value:{fmt}}"

    if colorize and isinstance(value, (int, float)):
        if value > 0:
            v_str = typer.style(v_str, fg=C.POSITIVE)
        elif value < 0:
            v_str = typer.style(v_str, fg=C.NEGATIVE)

    kv(label, v_str, label_width)


# ── Table Helpers ──────────────────────────────────────────────────────────

class Table:
    """Simple fixed-width table renderer that handles ANSI colors correctly.

    Usage:
        t = Table([("ID", 8), ("Name", 16), ("PnL", 10, "right")])
        t.header()
        t.row(["abc123", "my_strat", color_value(42.5)])
    """

    def __init__(self, columns: list[tuple[str, int] | tuple[str, int, str]]) -> None:
        self.columns: list[tuple[str, int, str]] = []
        for col in columns:
            if len(col) == 2:
                self.columns.append((col[0], col[1], "left"))
            else:
                self.columns.append((col[0], col[1], col[2]))

    def _pad(self, text: str, width: int, align: str) -> str:
        """Pad plain text to width, THEN apply any ANSI codes that are already in text."""
        # Strip ANSI to measure real length
        import re
        plain = re.sub(r'\033\[[0-9;]*m', '', text)
        padding = max(0, width - len(plain))
        if align == "right":
            return " " * padding + text
        return text + " " * padding

    def header(self) -> None:
        """Print bold header row and divider."""
        parts = [self._pad(c[0], c[1], c[2]) for c in self.columns]
        typer.echo()
        typer.echo("  " + bold("  ".join(parts)))
        total_w = sum(c[1] for c in self.columns) + (len(self.columns) - 1) * 2
        divider(total_w)

    def row(self, values: list[str]) -> None:
        """Print a data row. Values may contain ANSI color codes."""
        parts = []
        for i, val in enumerate(values):
            if i < len(self.columns):
                _, w, align = self.columns[i]
                parts.append(self._pad(str(val), w, align))
            else:
                parts.append(str(val))
        typer.echo("  " + "  ".join(parts))

    def footer(self) -> None:
        typer.echo()


# ── Card Helpers ───────────────────────────────────────────────────────────

def progress_bar(pct: int, width: int = 30) -> str:
    """Render a text progress bar."""
    filled = int(width * pct / 100) if pct else 0
    bar = typer.style("=" * filled, fg=C.POSITIVE) + dim("-" * (width - filled))
    return f"[{bar}] {pct}%"


def inline_progress(pct: int, status: str, elapsed: int, width: int = 20) -> str:
    """Render a single-line progress indicator for polling loops."""
    filled = int(width * pct / 100) if pct else 0
    bar = "=" * filled + "-" * (width - filled)
    sc = C.STATUS.get(status, C.NEUTRAL)
    return f"  [{bar}] {pct:>3d}%  {typer.style(status, fg=sc):>10s}  [{elapsed}s]"

File: cli/src/cli/style.rs
//! Shared CLI styling utilities — Rust equivalent of Python `_style.py`.
//!
//! Provides semantic colors, text helpers, table rendering, box-drawn reports,
//! and progress indicators for consistent terminal output across all commands.

use crossterm::style::{Color, Stylize};

/// Bright green (ANSI 92) — standard ANSI 32 green looks grey on light terminals.
pub const POS: Color = Color::AnsiValue(10);

// ── ANSI Helpers ─────────────────────────────────────────────────────────

/// Strip ANSI escape sequences from a string.
pub fn strip_ansi(s: &str) -> String {
    let mut result = String::with_capacity(s.len());
    let mut in_escape = false;
    for c in s.chars() {
        if in_escape {
            if c == 'm' {
                in_escape = false;
            }
            continue;
        }
        if c == '\x1b' {
            in_escape = true;
            continue;
        }
        result.push(c);
    }
    result
}

/// Get the visual (non-ANSI) length of a string.
pub fn ansi_len(s: &str) -> usize {
    strip_ansi(s).len()
}

/// Right-pad a string to a given visual width (ANSI-aware).
pub fn rpad(text: &str, width: usize) -> String {
    let visible = ansi_len(text);
    let padding = width.saturating_sub(visible);
    format!("{}{}", text, " ".repeat(padding))
}

// ── Text Helpers ─────────────────────────────────────────────────────────

pub fn bold(text: &str) -> String {
    format!("{}", text.bold())
}

pub fn dim(text: &str) -> String {
    format!("{}", text.dim())
}

pub fn accent(text: &str) -> String {
    format!("{}", text.cyan())
}

pub fn muted(text: &str) -> String {
    format!("{}", text.dark_grey())
}

/// Format and color a numeric value: green if positive, red if negative.
pub fn color_value(val: Option<f64>, fmt: &str) -> String {
    match val {
        None => muted("-"),
        Some(v) => {
            let s = match fmt {
                "+.2f" => format!("{:+.2}", v),
                ".2f" => format!("{:.2}", v),
                ".4f" => format!("{:.4}", v),
                "+.4f" => format!("{:+.4}", v),
                ".6f" => format!("{:.6}", v),
                "+.6f" => format!("{:+.6}", v),
                _ => format!("{:+.2}", v),
            };
            if v > 0.0 {
                format!("{}", s.with(POS))
            } else if v < 0.0 {
                format!("{}", s.red())
            } else {
                s
            }
        }
    }
}

/// Format a value without color.
pub fn format_value(val: Option<f64>, fmt: &str, suffix: &str) -> String {
    match val {
        None => muted("-"),
        Some(v) => {
            let s = match fmt {
                "+.2f" => format!("{:+.2}", v),
                ".2f" => format!("{:.2}", v),
                ".4f" => format!("{:.4}", v),
                _ => format!("{:.2}", v),
            };
            format!("{}{}", s, suffix)
        }
    }
}

/// Color a status string with status-specific color.
pub fn color_status(status: &str) -> String {
    match status {
        "completed" => format!("{}", status.with(POS).bold()),
        "running" => format!("{}", status.yellow().bold()),
        "queued" => format!("{}", status.cyan().bold()),
        "failed" | "error" => format!("{}", status.red().bold()),
        "cancelled" => format!("{}", status.magenta().bold()),
        _ => format!("{}", status.white().bold()),
    }
}

/// Render a colored [icon] badge for a status.
pub fn status_badge(status: &str) -> String {
    match status {
        "completed" => format!("{}", "[+]".with(POS)),
        "running" => format!("{}", "[~]".yellow()),
        "queued" => format!("{}", "[.]".cyan()),
        "failed" | "error" => format!("{}", "[x]".red()),
        "cancelled" => format!("{}", "[!]".magenta()),
        _ => "[?]".to_string(),
    }
}

// ── Layout Helpers ───────────────────────────────────────────────────────

/// Print a section header.
pub fn header(title: &str) {
    println!();
    println!("  {}", bold(title));
}

/// Print a horizontal divider.
pub fn divider(width: usize) {
    println!("  {}", "-".repeat(width));
}

/// Print a key-value pair with right-aligned label.
pub fn kv(label: &str, value: &str, label_width: usize) {
    println!("    {:>width$}: {}", label, value, width = label_width);
}

// ── Table ────────────────────────────────────────────────────────────────

pub enum Align {
    Left,
    Right,
}

struct TableColumn {
    name: String,
    width: usize,
    align: Align,
}

pub struct Table {
    columns: Vec<TableColumn>,
}

impl Table {
    pub fn new(cols: &[(&str, usize, &str)]) -> Self {
        let columns = cols
            .iter()
            .map(|(name, width, align)| TableColumn {
                name: name.to_string(),
                width: *width,
                align: if *align == "right" {
                    Align::Right
                } else {
                    Align::Left
                },
            })
            .collect();
        Self { columns }
    }

    fn pad(text: &str, width: usize, align: &Align) -> String {
        let visible = ansi_len(text);
        let padding = width.saturating_sub(visible);
        match align {
            Align::Left => format!("{}{}", text, " ".repeat(padding)),
            Align::Right => format!("{}{}", " ".repeat(padding), text),
        }
    }

    pub fn header(&self) {
        let parts: Vec<String> = self
            .columns
            .iter()
            .map(|c| Self::pad(&c.name, c.width, &c.align))
            .collect();
        println!();
        println!("  {}", bold(&parts.join("  ")));
        let total_w: usize =
            self.columns.iter().map(|c| c.width).sum::<usize>()
                + self.columns.len().saturating_sub(1) * 2;
        divider(total_w);
    }

    pub fn row(&self, values: &[&str]) {
        let parts: Vec<String> = values
            .iter()
            .enumerate()
            .map(|(i, val)| {
                if i < self.columns.len() {
                    Self::pad(val, self.columns[i].width, &self.columns[i].align)
                } else {
                    val.to_string()
                }
            })
            .collect();
        println!("  {}", parts.join("  "));
    }

    pub fn footer(&self) {
        println!();
    }
}

// ── Progress ─────────────────────────────────────────────────────────────

/// Render a text progress bar.
pub fn progress_bar(pct: u8, width: usize) -> String {
    let filled = (width as u16 * pct as u16 / 100) as usize;
    let empty = width.saturating_sub(filled);
    let bar_filled = format!("{}", "=".repeat(filled).with(POS));
    let bar_empty = dim(&"-".repeat(empty));
    format!("[{}{}] {}%", bar_filled, bar_empty, pct)
}

/// Render a single-line progress indicator for polling loops.
pub fn inline_progress(pct: u8, status: &str, elapsed: u64, width: usize) -> String {
    let filled = (width as u16 * pct as u16 / 100) as usize;
    let empty = width.saturating_sub(filled);
    let bar = format!("{}{}", "=".repeat(filled), "-".repeat(empty));
    format!(
        "  [{}] {:>3}%  {:>10}  [{}s]",
        bar,
        pct,
        color_status(status),
        elapsed,
    )
}

// ── Node Helpers ─────────────────────────────────────────────────────────

pub fn mode_label(mode: &str) -> String {
    match mode {
        "live" => format!("{}", "LIVE".yellow().bold()),
        _ => format!("{}", "SANDBOX".cyan().bold()),
    }
}

pub fn node_status_color(status: &str) -> String {
    match status {
        "running" => format!("{}", status.with(POS).bold()),
        "starting" => format!("{}", status.yellow().bold()),
        "stopped" => format!("{}", status.red().bold()),
        "error" => format!("{}", status.red().bold()),
        _ => format!("{}", status.white().bold()),
    }
}

pub fn node_badge(status: &str) -> String {
    match status {
        "running" => format!("{}", "[+]".with(POS)),
        "starting" => format!("{}", "[~]".yellow()),
        "stopped" => format!("{}", "[x]".red()),
        "error" => format!("{}", "[!]".red()),
        _ => "[?]".to_string(),
    }
}

// ── Box Drawing (for result reports) ─────────────────────────────────────

const BOX_H: &str = "\u{2500}";
const BOX_V: &str = "\u{2502}";
const BOX_TL: &str = "\u{250C}";
const BOX_TR: &str = "\u{2510}";
const BOX_BL: &str = "\u{2514}";
const BOX_BR: &str = "\u{2518}";
const BOX_LT: &str = "\u{251C}";
const BOX_RT: &str = "\u{2524}";

pub struct BoxReport {
    width: usize,
}

impl BoxReport {
    pub fn new(width: usize) -> Self {
        Self { width }
    }

    pub fn top(&self) {
        println!(
            "  {}{}{}",
            BOX_TL,
            BOX_H.repeat(self.width),
            BOX_TR,
        );
    }

    pub fn mid(&self) {
        println!(
            "  {}{}{}",
            BOX_LT,
            BOX_H.repeat(self.width),
            BOX_RT,
        );
    }

    pub fn bot(&self) {
        println!(
            "  {}{}{}",
            BOX_BL,
            BOX_H.repeat(self.width),
            BOX_BR,
        );
    }

    pub fn line(&self, text: &str) {
        let pad = self.width.saturating_sub(ansi_len(text) + 1);
        println!("  {} {}{}{}", BOX_V, text, " ".repeat(pad), BOX_V);
    }

    pub fn pair(&self, l1: &str, v1: &str, l2: &str, v2: &str) {
        let left = format!("{:>14}: {}", l1, rpad(v1, 12));
        let right = format!("{:>14}: {}", l2, v2);
        let combined = format!("{}  {}", left, right);
        self.line(&combined);
    }

    pub fn empty(&self) {
        self.line("");
    }
}
